Test/Graph/unit-graph.cc

```diff
@@ -40,26 +40,26 @@ TEST_CASE("Breadth-first search", "[graph]") {
   const auto input_tests = GetGraphInputs();
 
   for (const auto input : input_tests) {
-    netket::Graph graph(input);
+    auto graph = netket::make_graph(input);
     const auto data = input.dump();
 
     SECTION(
         "each node is visited at most once (and exactly once if the graph is "
         "connected) on " +
         data) {
-      for (int start = 0; start < graph.Nsites(); ++start) {
+      for (int start = 0; start < graph->Nsites(); ++start) {
```

Comment:
The only thing I don't like about this,overall, is that losing `Graph` and alikes we need to use pointers instead of objects. Calling methods with `->` is a pain, and it is especially painful to change all the methods calls throughout the code...

Reply:
AFAIK, the common practice is to do something like this
```cpp
auto pointer = make_*(...);
// in our case make_* functions indicate errors by throwing rather than returning nullptrs, so
assert(pointer != nullptr && "Bug! Postcondition violated");
auto& graph = *pointer;

... work with graph ...
```
Also, some people (mostly coming from C background I think) prefer to work directly with pointers.
Anyway, I don't think these extra lines of code are a good enough reason to drop this approach.

Reply:
Indeed it's a relatively minor problem, I just advocate a code-wide consistent use of `.`  instead of `->`  calls. I think the approach you proposed is good, we can certainly merge it (or maybe you need to do some cleanup first?)

Reply:
The other thing I was wondering: do you plan to keep the `hpp` and `cpp` implementations separate? I know it's good practice/can speed up compilation, but it's also a lot of duplication, I personally don't like the idea of having to read 2-3 different files to understand what's going on in a single class :) But again, it's most largely a matter taste... we can discuss about this

Reply:
> I think the approach you proposed is good, we can certainly merge it (or maybe you need to do some cleanup first?)

Python bindings are still a work in progress so there's not much to clean up now. Moving one or two functions, removing some dead code, perhaps, but I'd rather do it when we have more python classes. And I'd rather avoid too many forks in the code doing similar things.

> The other thing I was wondering: do you plan to keep the hpp and cpp implementations separate? I know it's good practice/can speed up compilation, but it's also a lot of duplication, I personally don't like the idea of having to read 2-3 different files to understand what's going on in a single class :) But again, it's most largely a matter taste... we can discuss about this

A big yes here :) It's not that much of a code duplication: only the signatures, and if you have a look at e.g. libc++ header files you'll notice that although almost everything is header-only, they add a synopsis at the top of the file. It's nice when you have to read through 200 lines of code rather than 2000 to understand the interface of the class. Kind of similar to why we have multiple header files.
As the project grows bigger (you already notice how annoying it'd be to replace `.` with `->`) wins in compile-time become more and more significant. Not so much for CI, but locally, when you've only changed a few lines in one function to test something out. Also, separate compilation allows the compiler to tell you more. E.g. while implementing this PR I've noticed several times that some `include`s were missing. It all works, because we include everything in `netket.hpp`, but it'll lead to a huge mess later. Another thing is detecting ODR violations (developers of header only libraries write tests specifically for that). Finally, SFINAE can lead to a lot of errors being unnoticed.
And as for the typical drawbacks like failing to inline something... We use virtual methods almost everywhere and they can't be inlined anyway, so I'm positive there's no performance overhead.
I was hoping to add separate compilation to Hilbert/* and then to Hamiltonian/* parts in the near future. So, please, let me know if you can't be persuaded to move to separate compilation or want to put it up for a vote etc. so that I don't spend time on something useless :)

Reply:
> Python bindings are still a work in progress so there's not much to clean up now. Moving one or two functions, removing some dead code, perhaps, but I'd rather do it when we have more python classes. And I'd rather avoid too many forks in the code doing similar things.

OK, let's organize the work then. This is going to be a massive change, touching potentially all the files in the repository. I can close my WIP PR (the changes you have done are not compatible upstream anymore I think), and create a new branch, `v2.0`.  Would you like to work a bit more by yourself on adding more Python classes and then push your changes to this new branch? I think that things become interesting when you start touching `Machine`.... 

> And as for the typical drawbacks like failing to inline something... We use virtual methods almost everywhere and they can't be inlined anyway, so I'm positive there's no performance overhead.

True, but my "dream" was indeed to eventually get rid of virtual methods in favour of Concepts. If we shift now to a compiled library instead of a header-only one, maybe it is going to be more painful to do the transition to concepts in the future? Can't we try to use precompiled headers or something intermediate?

Reply:
> I can close my WIP PR (the changes you have done are not compatible upstream anymore I think), and create a new branch, v2.0

By compatible you mean git-wise or user-experience-wise? Because if the users notice a difference, then I've most likely done something wrong and would be happy to fix it.

> Would you like to work a bit more by yourself on adding more Python classes and then push your changes to this new branch?

I should probably first fix the issues you pointer out in the comments... Also, we still need `make_hilbert` (and getting rid of `netket::Hilbert`), `make_hamiltonian` (and getting rid of `netket::Hamiltonian`), and `make_machine` (and getting rid of `netket::Machine`??) before we can wrap machines in python. So how about I start with these and you can create proper wrappers for graphs (like also handling the python list of lists to C++ vector of vectors conversions) in the meanwhile and then move to Hilbert spaces and Hamiltonians?

> True, but my "dream" was indeed to eventually get rid of virtual methods in favour of Concepts. If we shift now to a compiled library instead of a header-only one, maybe it is going to be more painful to do the transition to concepts in the future?

How long are you planning to wait for concepts? They're only coming in C++20 and NetKet is still using C++11 ;). And, more importantly, concepts aren't a replacement for inheritance. You'd have to make some pretty serious changes in the way things work, because it'll be impossible to have classes like `Graph` or `Machine`. Knowledge of which graph or machine you have becomes a compile-time thing, i.e. you have completely different code paths for different types and have to handle dispatching somewhere very close to main. And you'd have to create an amazing number of overloads for python stuff to work normally. Simply put, concepts make it easier to work with templates, they don't replace run-time polymorphism just like templates don't.

Reply:
> By compatible you mean git-wise or user-experience-wise? Because if the users notice a difference, then I've most likely done something wrong and would be happy to fix it.

I mean, In the last commits I had restructured the `Hilbert` class as something that must be specified in the input file, and is not automatically deduced (in some cases) from `Hamiltonian`. This was leading to a complicate initialization logic and also didn't make entirely sense, because the same Hamiltonian can be in principle defined on different Hilbert spaces (say Ising model for spin 1, or ising model for spin 1/2). So this breaks the current input file structure, since in most cases `Hilbert` was never specified in the input file. That's why I am saying it is maybe better to do a `v2.0` branch ? 

And agreed for the concepts thing. If there is no way to work with pre-compiled headers instead, then we can shift to independent compiled units.

Reply:
@twesterhout , let me know what you think about the plan. I can work on this full time this week, but will be very busy in the next weeks. So I'd prefer to get started as soon as possible without conflicts with what you plan to do. 

To get something simple and easy to maintain, the plan is to get rid of JSON constructors, ok `make_hypercube` etc, and of `Graph` and alikes. The real work to do is to set up normal constructors for the classes we have, and write the pybind11 bindings for those, as derived classes from the abstract ones. This will imply that we need to rewrite the unit tests as well, starting from Graph and then moving on. It's quite some work, but it's scalable and in the future will be much easier to maintain. Just let me know if I can get started on this tomorrow.

Reply:
>  If there is no way to work with pre-compiled headers instead, then we can shift to independent compiled units.

I'll be honest that I don't have much experience with `.pch` files. They have some issues of their own like being usable in but one TU I think. And while they might help with compile times, they definitely don't solve all the other issues I mentioned.

> I mean, In the last commits I had restructured the Hilbert class as something that must be specified in the input file, and is not automatically deduced (in some cases) from Hamiltonian. This was leading to a complicate initialization logic and also didn't make entirely sense, because the same Hamiltonian can be in principle defined on different Hilbert spaces (say Ising model for spin 1, or ising model for spin 1/2). So this breaks the current input file structure, since in most cases Hilbert was never specified in the input file. That's why I am saying it is maybe better to do a v2.0 branch ?

I see. Is sounds to me that this can be handled in a single branch somewhere in ground_state.hpp where different parts (graph, Hamiltonian, machine, etc.) are constructed. Or am I missing something? We could add a message saying that the feature of deducing Hilbert space automatically is deprecated and will be removed after X releases.
On the other hand, if new constructors etc. are taken into account, we get a major rewrite of the code base, so it does make sense to call it v2 even if no user code is broken... Different projects do it differently, so perhaps this decision should be put up for a vote or something? It doesn't slow us down a bit because it's just the name of the branch that changes.


About dropping support for JSON input. NetKet is quite (read very) far from being usable as a C++ library and while it might be a good long time goal, I'd vote against focusing on it now.
Also, you said nothing about what'd happen to the `netket` executable. Are you considering dropping it as well? For example, when I'm testing something related to NetKet, I almost never use the python scripts from the Tutorial folder. Instead, I just write a JSON input file. It's also easier to generate a few JSON files in bash using simple variable substitutions than adding command line arguments support to python scripts and then calling them to generateJSON for me. This is useful for submitting multiple jobs at once to clusters. Finally, the overhead (code-wise) of keeping JSON support is not high at all: just a couple `make_*` functions and a main program. I'd even sign up for maintaining this part if we had proper constructors handling invalid input correctly.

All in all, yes, I think there's nothing stopping you from getting started on this right away. One thing: maybe we should push this branch to netket/something so the conversation and PRs become visible to everyone?

Reply:
@twesterhout , let me clarify, in my proposal (that I have sketched in the branch v2.0) : 
1. the `netket` executable would disappear 
2. this entails that c++ unit tests should be written in terms of the single classes and not in terms of JSON 
3. I think that the most common use pattern is not to write directly the JSON file, but use the python scripts we have in the folder. That's why moving to a (much more usable and flexibly) python library there would be certainly good
4. Notice that if you have a full-fledged python libraries you can also very easily submit multiple jobs (your approach to submitting multiple jobs is neat, but I am not aware of many people doing that :) ) 
5. Yes, we have to consider the overhead of maintaining the JSON support. In the branch v2.0 I have shown that it is possible to keep them alive together. However I am against it, because I think it's a headache to maintain all the possible ways of constructing classes... 

I agree with you it's good to move the v2.0 branch to the main netket repo and gather also other opinions on how to proceed. It might well be that I have a distorted vision of what people expect from this library and that JSON support is essential for them, who knows.